mlflow/tracking/client.py

```diff
@@ -433,6 +440,217 @@ def get_trace_info(self, trace_id: str) -> TraceInfo:
+
+        Returns:
+            An MLflowSpanWrapper object representing the span.
+
+        Example:
+
+        .. code-block:: python
+
+                from mlflow.tracking import MlflowClient
+
+                client = MlflowClient()
+
+                span = client.start_trace("my_trace")
+
+                x = 2
+
+                # Create a child span
+                child_span = client.start_span(
+                    "child_span",
+                    trace_id=span.trace_id,
+                    parent_span_id=span.id,
+                    inputs={"x": 2},
+                )
+
+                y = x**2
+
+                client.end_span(
+                    trace_id=child_span.trace_id,
+                    span_id=child_span.span_id,
+                    attributes={"factor": 2},
+                    outputs={"y": y},
+                )
+
+                client.end_trace(trace_id)
+        """
+        if not parent_span_id:
+            raise MlflowException(
+                "start_span() must be called with an explicit parent_span_id."
+                "If you haven't start any span yet, use MLflowClient().start_trace() "
+                "to start a new trace and root span.",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+        trace_manager = InMemoryTraceManager.get_instance()
+        span = trace_manager.start_detached_span(
+            name=name,
+            trace_id=trace_id,
+            parent_span_id=parent_span_id,
+            span_type=span_type,
+        )
+
+        if attributes:
+            span.set_attributes(attributes)
+        if inputs:
+            span.set_inputs(inputs)
+
+        return span
+
+    def end_span(
+        self,
+        trace_id: str,
+        span_id: str,
+        outputs: Optional[Dict[str, Any]] = None,
+        attributes: Optional[Dict[str, Any]] = None,
+        status: Status = Status(StatusCode.OK, ""),
+    ):
+        """
+        End the span with the given trace ID and span ID.
+
+        Args:
+            trace_id: The ID of the trace to end.
+            span_id: The ID of the span to end.
+            outputs: Outputs to set on the span.
+            attributes: A dictionary of attributes to set on the span. If the span already has
+                attributes, the new attributes will be merged with the existing ones. If the same
+                key already exists, the new value will overwrite the old one.
+            status: The status of the span. The default status is OK.
+        """
+        trace_manager = InMemoryTraceManager.get_instance()
+        span = trace_manager.get_span_from_id(trace_id, span_id)
+
+        if span is None:
+            _logger.warning(f"Span with ID {span_id} not found in trace with ID {trace_id}.")
+            return
+
+        if attributes:
+            span.set_attributes(attributes or {})
+        if outputs:
+            span.set_outputs(outputs)
+        span.set_status(status)
+
```

Comment:
Do we need an `info` here if the span is root_span, we say `trace {trace_id} is ended` to avoid user from running `end_trace` again? I'm not sure if trace operations will confuse users as start_trace must be called while end_span(root_span) is the same as end_trace

Reply:
I agree that it might be confusing to have trace operations that actually does span, but I doubt `info` is helpful as they are not rendered in Databricks notebook, and even in notebook there will be a lot of `info` level logs will be emitted. I think we can see users feedback and consider more concrete solutions afterwards.